File: automation-test-playwright/pages/admin/AdminReportPage.ts
import { AdminShellPage } from "../core/AdminShellPage";

export class AdminReportPage extends AdminShellPage {
  protected readonly path = "/admin/report";

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/report/);
    await this.waitForVisible(this.page.locator(".year-select, select[name='year']"));
    await this.waitForVisible(this.page.locator("#monthlyRevenueChart"));
  }

  async waitForOverviewVisible(): Promise<void> {
    await this.waitForLocatorCount(this.page.locator(".stat-card"), 3);
    await this.waitForVisible(this.page.locator("#monthlyRevenueChart"));
    await this.waitForVisible(this.page.locator("#rentGrowthChart"));
    await this.waitForVisible(this.page.locator("#yearlyRevenueChart"));
    await this.waitForVisible(this.page.locator("#propertyTypeChart"));
    await this.waitForVisible(this.page.locator("#topBuildingChart"));
  }

  async selectYear(value: string): Promise<void> {
    await this.firstVisible(this.page.locator(".year-select, select[name='year']")).selectOption(value);
  }

  async availableYears(): Promise<string[]> {
    return this.firstVisible(this.page.locator(".year-select, select[name='year']"))
      .locator("option")
      .evaluateAll((options) => options.map((option) => (option as HTMLOptionElement).value));
  }

  async submitYearFilter(): Promise<void> {
    const submitButton = this.firstVisible(this.page.locator("form.year-selector-form button, button[type='submit']"));
    if (await submitButton.count()) {
      await submitButton.click();
    }
  }

  async waitForYearSelected(value: string): Promise<void> {
    await this.waitForLocatorValue(this.firstVisible(this.page.locator(".year-select, select[name='year']")), value);
  }

  /**
   * Triggers the print action.
   */
  async triggerPrint(): Promise<void> {
    await this.page.locator('button[onclick*="window.print"]').click();
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminReportPage.java
package org.suntower.pages.admin;

import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.suntower.core.BasePage;

public class AdminReportPage extends BasePage {
  private final By yearSelect = css(".year-select, select[name='year']");

  public AdminReportPage(WebDriver driver) {
    super(driver);
  }

  public void open() {
    visit("/admin/report");
  }

  public void waitForLoaded() {
    waitForUrlContains("/admin/report");
    waitForVisible(yearSelect);
    waitForVisible(css("#monthlyRevenueChart"));
  }

  public void waitForOverviewVisible() {
    waitForCondition(() -> driver.findElements(css(".stat-card")).size() >= 3, "Report stat cards did not load.");
    waitForVisible(css("#monthlyRevenueChart"));
    waitForVisible(css("#rentGrowthChart"));
    waitForVisible(css("#yearlyRevenueChart"));
    waitForVisible(css("#propertyTypeChart"));
    waitForVisible(css("#topBuildingChart"));
  }

  public List<String> availableYears() {
    return visible(yearSelect).findElements(By.cssSelector("option")).stream().map(option -> option.getAttribute("value")).toList();
  }

  public void selectYear(String year) {
    setInputValue(yearSelect, year);
  }

  public void waitForYearSelected(String year) {
    waitForValue(yearSelect, year);
  }

  public void triggerPrint() {
    ((JavascriptExecutor) driver).executeScript("window.__printTriggered = false; window.print = function(){ window.__printTriggered = true; };");
    click(firstVisible(css("button[onclick*='window.print']")));
  }

  public boolean wasPrintTriggered() {
    return Boolean.TRUE.equals(((JavascriptExecutor) driver).executeScript("return window.__printTriggered === true;"));
  }
}
